### ISSUE TITLE
ajax on front - atksample_form

### ISSUE THREAD

**Issue by RobertMassart** [ibelar/atk-wordpress-demo#4]
I moved this into demo to see if it would work:

<?php
/**
 * Display a simple registration form using a shortcode.
 */

/**
 * insert this code "[atksample-form]" into a page to get the register form display in Wp front end.
 */
namespace atkdemo\shortcodes;

use atkwp\components\ShortcodeComponent;

class FormShortcode extends ShortcodeComponent
{
    public function init()
    {
        parent::init();

        $form = $this->add('Form');
        $form->addField('email');
        $form->onSubmit(function ($form) {
            if (!$form->model['email'] || empty($form->model['email'])) {
                return $form->error('email', 'Please add valid email');
            }
            // implement subscribe here
            return $form->success('Subscribed '.$form->model['email'].' to newsletter.');
        });

        $form->buttonSave->set('Subscribe');
        $form->buttonSave->icon = 'mail';
    }
}

but same error as sample : 

jquery.js?ver=1.12.4-wp:4 POST http://dev.peernet.com/mytestatk/admin-ajax.php?action=atkd&atkwp=register&atkwp-count=1&_ajax_nonce=09903e4043&atkd-1_formshortcode_form_submit=ajax&__atk_callback=1 404 (Not Found)
send @ jquery.js?ver=1.12.4-wp:4
ajax @ jquery.js?ver=1.12.4-wp:4
xhr @ semantic.min.js?ver=2.6.4:11
request @ semantic.min.js?ver=2.6.4:11
query @ semantic.min.js?ver=2.6.4:11
trigger @ semantic.min.js?ver=2.6.4:11
dispatch @ jquery.js?ver=1.12.4-wp:3
r.handle @ jquery.js?ver=1.12.4-wp:3
trigger @ jquery.js?ver=1.12.4-wp:3
(anonymous) @ jquery.js?ver=1.12.4-wp:3
each @ jquery.js?ver=1.12.4-wp:2
each @ jquery.js?ver=1.12.4-wp:2
trigger @ jquery.js?ver=1.12.4-wp:3
n.fn.<computed> @ jquery.js?ver=1.12.4-wp:4
submit @ semantic.min.js?ver=2.6.4:11
invoke @ semantic.min.js?ver=2.6.4:11
initialize @ semantic.min.js?ver=2.6.4:11
(anonymous) @ semantic.min.js?ver=2.6.4:11
each @ jquery.js?ver=1.12.4-wp:2
each @ jquery.js?ver=1.12.4-wp:2
E.fn.form @ semantic.min.js?ver=2.6.4:11
(anonymous) @ (index):414
dispatch @ jquery.js?ver=1.12.4-wp:3
r.handle @ jquery.js?ver=1.12.4-wp:3
semantic.min.js?ver=2.6.4:11 API: Server gave an error: error admin-ajax.php?action=atkd&atkwp=register&atkwp-count=1&_ajax_nonce=09903e4043&atkd-1_formshortcode_form_submit=ajax&__atk_callback=1
error @ semantic.min.js?ver=2.6.4:11
fail @ semantic.min.js?ver=2.6.4:11
i @ jquery.js?ver=1.12.4-wp:2
fireWith @ jquery.js?ver=1.12.4-wp:2
(anonymous) @ semantic.min.js?ver=2.6.4:11
setTimeout (async)
fail @ semantic.min.js?ver=2.6.4:11
i @ jquery.js?ver=1.12.4-wp:2
fireWith @ jquery.js?ver=1.12.4-wp:2
x @ jquery.js?ver=1.12.4-wp:4
c @ jquery.js?ver=1.12.4-wp:4
XMLHttpRequest.send (async)
send @ jquery.js?ver=1.12.4-wp:4
ajax @ jquery.js?ver=1.12.4-wp:4
xhr @ semantic.min.js?ver=2.6.4:11
request @ semantic.min.js?ver=2.6.4:11
query @ semantic.min.js?ver=2.6.4:11
trigger @ semantic.min.js?ver=2.6.4:11
dispatch @ jquery.js?ver=1.12.4-wp:3
r.handle @ jquery.js?ver=1.12.4-wp:3
trigger @ jquery.js?ver=1.12.4-wp:3
(anonymous) @ jquery.js?ver=1.12.4-wp:3
each @ jquery.js?ver=1.12.4-wp:2
each @ jquery.js?ver=1.12.4-wp:2
trigger @ jquery.js?ver=1.12.4-wp:3
n.fn.<computed> @ jquery.js?ver=1.12.4-wp:4
submit @ semantic.min.js?ver=2.6.4:11
invoke @ semantic.min.js?ver=2.6.4:11
initialize @ semantic.min.js?ver=2.6.4:11
(anonymous) @ semantic.min.js?ver=2.6.4:11
each @ jquery.js?ver=1.12.4-wp:2
each @ jquery.js?ver=1.12.4-wp:2
E.fn.form @ semantic.min.js?ver=2.6.4:11
(anonymous) @ (index):414
dispatch @ jquery.js?ver=1.12.4-wp:3
r.handle @ jquery.js?ver=1.12.4-wp:3
semantic.min.js?ver=2.6.4:11 Transition: Element is no longer attached to DOM. Unable to animate.  Use silent setting to surpress this warning in production. fade in <div class=​"ui dimmer modals page">​…​</div>​
error @ semantic.min.js?ver=2.6.4:11
animate @ semantic.min.js?ver=2.6.4:11
initialize @ semantic.min.js?ver=2.6.4:11
(anonymous) @ semantic.min.js?ver=2.6.4:11
each @ jquery.js?ver=1.12.4-wp:2
each @ jquery.js?ver=1.12.4-wp:2
C.fn.transition @ semantic.min.js?ver=2.6.4:11
show @ semantic.min.js?ver=2.6.4:11
show @ semantic.min.js?ver=2.6.4:11
invoke @ semantic.min.js?ver=2.6.4:11
(anonymous) @ semantic.min.js?ver=2.6.4:11
each @ jquery.js?ver=1.12.4-wp:2
each @ jquery.js?ver=1.12.4-wp:2
k.fn.dimmer @ semantic.min.js?ver=2.6.4:11
showDimmer @ semantic.min.js?ver=2.6.4:11
showModal @ semantic.min.js?ver=2.6.4:11
show @ semantic.min.js?ver=2.6.4:11
invoke @ semantic.min.js?ver=2.6.4:11
(anonymous) @ semantic.min.js?ver=2.6.4:11
each @ jquery.js?ver=1.12.4-wp:2
each @ jquery.js?ver=1.12.4-wp:2
I.fn.modal @ semantic.min.js?ver=2.6.4:11
value @ atkjs-ui.min.js?ver=1.6:1
value @ atkjs-ui.min.js?ver=1.6:1
fail @ semantic.min.js?ver=2.6.4:11
i @ jquery.js?ver=1.12.4-wp:2
fireWith @ jquery.js?ver=1.12.4-wp:2
(anonymous) @ semantic.min.js?ver=2.6.4:11
setTimeout (async)
fail @ semantic.min.js?ver=2.6.4:11
i @ jquery.js?ver=1.12.4-wp:2
fireWith @ jquery.js?ver=1.12.4-wp:2
x @ jquery.js?ver=1.12.4-wp:4
c @ jquery.js?ver=1.12.4-wp:4
XMLHttpRequest.send (async)
send @ jquery.js?ver=1.12.4-wp:4
ajax @ jquery.js?ver=1.12.4-wp:4
xhr @ semantic.min.js?ver=2.6.4:11
request @ semantic.min.js?ver=2.6.4:11
query @ semantic.min.js?ver=2.6.4:11
trigger @ semantic.min.js?ver=2.6.4:11
dispatch @ jquery.js?ver=1.12.4-wp:3
r.handle @ jquery.js?ver=1.12.4-wp:3
trigger @ jquery.js?ver=1.12.4-wp:3
(anonymous) @ jquery.js?ver=1.12.4-wp:3
each @ jquery.js?ver=1.12.4-wp:2
each @ jquery.js?ver=1.12.4-wp:2
trigger @ jquery.js?ver=1.12.4-wp:3
n.fn.<computed> @ jquery.js?ver=1.12.4-wp:4
submit @ semantic.min.js?ver=2.6.4:11
invoke @ semantic.min.js?ver=2.6.4:11
initialize @ semantic.min.js?ver=2.6.4:11
(anonymous) @ semantic.min.js?ver=2.6.4:11
each @ jquery.js?ver=1.12.4-wp:2
each @ jquery.js?ver=1.12.4-wp:2
E.fn.form @ semantic.min.js?ver=2.6.4:11
(anonymous) @ (index):414
dispatch @ jquery.js?ver=1.12.4-wp:3
r.handle @ jquery.js?ver=1.12.4-wp:3
semantic.min.js?ver=2.6.4:11 Transition: Element is no longer attached to DOM. Unable to animate.  Use silent setting to surpress this warning in production. scale in <div class=​"ui modal" style=​"padding:​ 1em;​">​…​</div>​
error @ semantic.min.js?ver=2.6.4:11
animate @ semantic.min.js?ver=2.6.4:11
initialize @ semantic.min.js?ver=2.6.4:11
(anonymous) @ semantic.min.js?ver=2.6.4:11
each @ jquery.js?ver=1.12.4-wp:2
each @ jquery.js?ver=1.12.4-wp:2
C.fn.transition @ semantic.min.js?ver=2.6.4:11
showModal @ semantic.min.js?ver=2.6.4:11
show @ semantic.min.js?ver=2.6.4:11
invoke @ semantic.min.js?ver=2.6.4:11
(anonymous) @ semantic.min.js?ver=2.6.4:11
each @ jquery.js?ver=1.12.4-wp:2
each @ jquery.js?ver=1.12.4-wp:2
I.fn.modal @ semantic.min.js?ver=2.6.4:11
value @ atkjs-ui.min.js?ver=1.6:1
value @ atkjs-ui.min.js?ver=1.6:1
fail @ semantic.min.js?ver=2.6.4:11
i @ jquery.js?ver=1.12.4-wp:2
fireWith @ jquery.js?ver=1.12.4-wp:2
(anonymous) @ semantic.min.js?ver=2.6.4:11
setTimeout (async)
fail @ semantic.min.js?ver=2.6.4:11
i @ jquery.js?ver=1.12.4-wp:2
fireWith @ jquery.js?ver=1.12.4-wp:2
x @ jquery.js?ver=1.12.4-wp:4
c @ jquery.js?ver=1.12.4-wp:4
XMLHttpRequest.send (async)
send @ jquery.js?ver=1.12.4-wp:4
ajax @ jquery.js?ver=1.12.4-wp:4
xhr @ semantic.min.js?ver=2.6.4:11
request @ semantic.min.js?ver=2.6.4:11
query @ semantic.min.js?ver=2.6.4:11
trigger @ semantic.min.js?ver=2.6.4:11
dispatch @ jquery.js?ver=1.12.4-wp:3
r.handle @ jquery.js?ver=1.12.4-wp:3
trigger @ jquery.js?ver=1.12.4-wp:3
(anonymous) @ jquery.js?ver=1.12.4-wp:3
each @ jquery.js?ver=1.12.4-wp:2
each @ jquery.js?ver=1.12.4-wp:2
trigger @ jquery.js?ver=1.12.4-wp:3
n.fn.<computed> @ jquery.js?ver=1.12.4-wp:4
submit @ semantic.min.js?ver=2.6.4:11
invoke @ semantic.min.js?ver=2.6.4:11
initialize @ semantic.min.js?ver=2.6.4:11
(anonymous) @ semantic.min.js?ver=2.6.4:11
each @ jquery.js?ver=1.12.4-wp:2
each @ jquery.js?ver=1.12.4-wp:2
E.fn.form @ semantic.min.js?ver=2.6.4:11
(anonymous) @ (index):414
dispatch @ jquery.js?ver=1.12.4-wp:3
r.handle @ jquery.js?ver=1.12.4-wp:3

I think the ajax url is wrong on the front side but not sure.

**Comment by RobertMassart** [comment]
Had to fix this function for wordpress ajax on front end to have the correct admin-ajax URL

    /**
     * Return url.
     *
     * @param array $page
     * @param bool  $hasRequestUri
     * @param array $extraArgs
     *
     * @return array|null|string
     */
    public function jsUrl($page = [], $hasRequestUri = false, $extraArgs = [])
    {
        if (is_string($page)) {
            return $page;
        }

        $wpPage = 'admin-ajax';

        //if running front end set url for ajax.
        if (!WpUtil::isAdmin()) {
            $this->page = WpUtil::getBaseAdminUrl().'admin-ajax';
        }

        // RJM - Had add this to get URL correct for front end
        $wpPage = WpUtil::getBaseAdminUrl().'admin-ajax';
